src/main/java/com/google/devtools/build/skydoc/rendering/MarkdownUtil.java

```diff
@@ -46,55 +47,131 @@ public MarkdownUtil(String extensionBzlFile) {
   }
 
   /**
-   * Return a string that formats the input string so it is displayable in a markdown table cell.
-   * This performs the following operations:
+   * Formats the input string so that it is displayable in a Markdown table cell. This performs the
+   * following operations:
    *
    * <ul>
    *   <li>Trims the string of leading/trailing whitespace.
-   *   <li>Transforms the string using {@link #htmlEscape}.
-   *   <li>Transforms multline code (```) tags into preformatted code HTML tags.
-   *   <li>Transforms single-tick code (`) tags into code HTML tags.
-   *   <li>Transforms 'new paraphgraph' patterns (two or more sequential newline characters) into
-   *       line break HTML tags.
-   *   <li>Turns lingering new line tags into spaces (as they generally indicate intended line wrap.
+   *   <li>Escapes pipe characters ({@code |}) as {@code \|}.
+   *   <li>Transforms Markdown code blocks ({@code ```}) into HTML preformatted code blocks, and
+   *       transforms newlines within those code blocks into character entities
+   *   <li>Transforms remaining 'new paragraph' patterns (two or more sequential newline characters)
+   *       into line break HTML tags.
+   *   <li>Turns remaining newlines into spaces (as they generally indicate intended line wrap).
    * </ul>
+   *
+   * TODO(https://github.com/bazelbuild/stardoc/issues/118): also format Markdown lists as HTML.
    */
-  public String markdownCellFormat(String docString) {
-    String resultString = htmlEscape(docString.trim());
+  public static String markdownCellFormat(String docString) {
+    return new MarkdownCellFormatter(docString).format();
+  }
 
-    resultString = replaceWithTag(resultString, "```", "<pre><code>", "</code></pre>");
-    resultString = replaceWithTag(resultString, "`", "<code>", "</code>");
+  // See https://github.github.com/gfm
+  private static final class MarkdownCellFormatter {
+    private final ImmutableList<String> lines;
```

Comment:
Please add one-line comments to doc the invariant for these fields. (Especially that currentLine is 0-indexed.)